frame/staking/src/lib.rs

```diff
@@ -1647,19 +1950,120 @@ impl<T: Trait> Module<T> {
+			let exposure = <Module<T> as Store>::Stakers::get(validator);
+			current_total_stake += exposure.total;
+			<Module<T> as Store>::ErasStakers::insert(current_era, validator, &exposure);
+
+			let mut exposure_clipped = exposure;
+			let clipped_max_len = T::MaxNominatorRewardedPerValidator::get() as usize;
+			if exposure_clipped.others.len() > clipped_max_len {
+				exposure_clipped.others.sort_unstable_by(|a, b| a.value.cmp(&b.value).reverse());
+				exposure_clipped.others.truncate(clipped_max_len);
+			}
+			<Module<T> as Store>::ErasStakersClipped::insert(current_era, validator, exposure_clipped);
+
+			let pref = <Module<T> as Store>::Validators::get(validator);
+			<Module<T> as Store>::ErasValidatorPrefs::insert(current_era, validator, pref);
+		}
+		<Module<T> as Store>::ErasTotalStake::insert(current_era, current_total_stake);
+
+		let points = <Module<T> as Store>::CurrentEraPointsEarned::get();
+		<Module<T> as Store>::ErasRewardPoints::insert(current_era, EraRewardPoints {
+			total: points.total,
+			individual: current_elected.iter().cloned().zip(points.individual.iter().cloned()).collect(),
+		});
+
+		let res = <Module<T> as Store>::Ledger::translate_values(
+			|old: OldStakingLedger<T::AccountId, BalanceOf<T>>| StakingLedger {
+				stash: old.stash,
+				total: old.total,
+				active: old.active,
+				unlocking: old.unlocking,
+				last_reward: None,
+			}
+		);
+		if let Err(e) = res {
+			frame_support::print("Encountered error in migration of Staking::Ledger map.");
+			frame_support::print("The number of removed key/value is:");
+			frame_support::print(e);
+		}
+
+
+		// Kill old storages
+		<Module<T> as Store>::Stakers::remove_all();
+		<Module<T> as Store>::SlotStake::kill();
+		<Module<T> as Store>::CurrentElected::kill();
+		<Module<T> as Store>::CurrentEraStart::kill();
+		<Module<T> as Store>::CurrentEraStartSessionIndex::kill();
+		<Module<T> as Store>::CurrentEraPointsEarned::kill();
+	}
 }
 
+/// In this implementation `new_session(session)` must be called before `end_session(session-1)`
+/// i.e. the new session must be planned before the ending of the previous session.
+///
+/// Once the first new_session is planned, all session must start and then end in order, though
+/// some session can lag in between the newest session planned and the latest session started.
 impl<T: Trait> pallet_session::SessionManager<T::AccountId> for Module<T> {
 	fn new_session(new_index: SessionIndex) -> Option<Vec<T::AccountId>> {
 		Self::ensure_storage_upgraded();
-		if new_index == 0 {
-			return Self::initial_session();
-		}
-		Self::new_session(new_index - 1)
+		Self::new_session(new_index)
+	}
+	fn start_session(start_index: SessionIndex) {
+		Self::start_session(start_index)
+	}
+	fn end_session(end_index: SessionIndex) {
+		Self::end_session(end_index)
 	}
-	fn end_session(_end_index: SessionIndex) {}
 }
 
+/// This implementation has the same constrains as the implementation of
+/// `pallet_session::SessionManager`.
```

Comment:
This session manager implementation is still using a deprecated storage item.

```
let exposure = <Stakers<T>>::get(&v);
```

Can we remove all the deprecated code from this PR, and hard code some of the storage names in the migration function? Would help mitigate issues like this.

Reply:
I'm on it, I'll make deprecated storage out of the decl_storage